[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #32816 [quant][graphmode] Add add_relu pattern in skip values
* #32814 [quant][graphmode] Skip quantizing input and output in matched module
* #33277 [WIP][quant][graphmode] Observing input/output values in callsite
* #33649 [quant][graphmode][refactor] Factor out getInvokedMethod
* #33276 [quant][graphmode][refactor] Move check for weight outside of insertObserverFor
* #33275 [quant][graphmode][refactor] Move values_to_skip check inside valueNeedsToBeQuantized
* #33274 [quant][graphmode][refactor] Simplify signature for insertObserverFor
* #33273 [quant][graphmode][refactor] Checks for bias and weight
* #33173 [quant][graphmode] Add a pass to handle ops that doesn't require observation
* #33531 [quant][graphmode] Replicate dequantize nodes
* #33172 [quant][graphmode][refactor] Factor out insertDequantCall
* #33171 [quant][graphmode] refactor nodeQuantizable
* #32813 [quant][graphmode][refactor] Separate preprocess step for insertObserver
* #32812 [quant][graphmode][refactor] Change signature of getModuleAccessPath
* #32809 [quant][graphmode][refactor] Move the check for qconfig inside insertObserver call
* **#32379 [quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes**

Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D20078719](https://our.internmc.facebook.com/intern/diff/D20078719)

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 8dce821e:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8dce821e8ff37b52f4d473ccb5e447b26eef7749).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4297171) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=88060979) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=88060979))
<details>
<summary>
<code>Jan 17 18:53:30 RuntimeError: test_jit failed!</code>
</summary>

```
Jan 17 18:53:29 Ran 2341 tests in 105.725s 
Jan 17 18:53:29  
Jan 17 18:53:29 FAILED (errors=3, skipped=68, expected failures=1) 
Jan 17 18:53:29  
Jan 17 18:53:29 Generating XML reports... 
Jan 17 18:53:30 Traceback (most recent call last): 
Jan 17 18:53:30   File "test/run_test.py", line 456, in <module> 
Jan 17 18:53:30     main() 
Jan 17 18:53:30   File "test/run_test.py", line 449, in main 
Jan 17 18:53:30     raise RuntimeError(message) 
Jan 17 18:53:30 RuntimeError: test_jit failed! 
Jan 17 18:53:30 + cleanup 
Jan 17 18:53:30 + retcode=1 
Jan 17 18:53:30 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1488,96 +1500,109 @@ graph(%self, %x):
+    auto w_b = item.second;
+    conv.setattr("weight", std::get<0>(w_b));
+    // registering a parameter field (not attribute field)
+    conv.register_parameter("bias", std::get<1>(w_b), false);
```
What happens if `bias` already exists as a parameter?

jerryzh168 replied:
That's fine, we'll overwrite it

jerryzh168 replied:
this does two things, if parameter bias already exist, we'll overwrite, if not(we deleted bias attribute in analyze step), we'll register a bias parameter

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1627,8 +1652,11 @@ void QuantFusion(std::shared_ptr<Graph>& graph) {
+  script::Module m = module.clone();
+  h.analyze(m);
+  h.transform();
+  h.cleanup();
```
While I like the split into analysis and transformation, we need to add comments what exactly is happening when we're doing this. Please describe what data the analysis collects/verifies, what properties transformation expects, what we're doing in the cleanup. Without this, the code is not much better than
```
doStuff();
doMoreStuff();
doExtraStuff();
```

jerryzh168 replied:
Sure, will add more comments

ZolotukhinM reviewed (commented):
Please add a more descriptive commit message - what does it mean to support shared ClassTypes? What do shared ClassTypes even mean? What changed compared to the previous implementation and why this change was needed? Note that these commit messages are not for you or me, who has some context, but rather for a person who will try to understand what's going on some time in future.

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Supporting folding conv bn in shared ClassTypes

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM reviewed (commented):
Overall it seems reasonable, but I have remarks, please find them inline!

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1405,7 +1405,32 @@ static bool hastensor(script::Module& m, const char* name) {
+   * actually performing these modifications.
+   * Also we'll remove the bias attribute if it is non-Parameter
+   * and None, since in the transform step we'll add a bias Parameter
+   * of type Tensor
```
Oh, this comment is now outdated, I'll update

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1488,96 +1511,95 @@ graph(%self, %x):
+      Graph* g = method.graph().get();
+      // not successful insert means it already exists in the set
+      bool is_folded_graph = !folded_graph_.insert(g).second;
+      for (const Match& match : matches) {
```
we need to extract convbn parameters from this module instance as well, I don't think that can be achieved without matching the module instance Value in the graph.

jerryzh168 replied:
right, we can store these

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1488,96 +1508,109 @@ graph(%self, %x):
+
+          ConvBNParameters params;
+          if (!tryExtractingConvBNParameters(
+                  conv_submodule, bn_submodule, params)) {
```
Where are the extracted params used?

jerryzh168 replied:
This is to filter out invalid matches

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1511,14 +1510,16 @@ def forward(self, x):
                 self.bn = torch.nn.BatchNorm2d(num_features=20)
+                # to make sure new bias is not zero
+                self.bn.eps = 0.0027
+                self.bn.bias = torch.nn.Parameter(torch.rand([20]))
```
Can we make the test below an exact check (L 1550)?

raghuramank100 reviewed (changes requested):
Please see comments

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] FoldConvBatchNorm2d support shared ClassTypes"


Summary:
Folding Conv2d - BatchNorm2d modules means recalculate the weight and bias of Conv2d module by incorproating the parameters
of BatchNorm2d, and also change the method calls to calling only forward of Conv2d module, this involves change of both module
types and graph because the bias of Conv2d is a parameter when it has value and is an attribute when it is
None(since JIT code has assumption of prameter being Tensor in multiple places), therefore
we'll need to remove the bias attribute when it is None and add a bias attribute later. Since ClassType might be shared, we separate
remove and add in separate steps and also keep track of the processed graph to avoid modifying the graph and type multiple times.
However we'll have to record the slot index of bias as well so we can replay the slot removal on other instances of Conv2d module.

Test Plan:
tbd

Reviewers:
mvz, suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
